### PR TITLE
Build Linux, MacOS and Windows platform in Github Actions CI

### PR DESCRIPTION
**Pull request by DeinAlptraum**
This is a first attempt at a Github Actions CI. It currently successfully builds for Windows (32 btit), MacOS (Intel) and Linux (64bit).

There's still a few problems and open questions though:
- the Windows build and the portable upload runs for me, but the produced installer does not. Despite following the exact same steps for the installer as the current CI, it immediately crashes with the following error:
scr
I do not know what causes this error (no, "run as admin" doesn't fix this) and googling it did not really help either. Since I have no experience with NSIS, maybe someone here knows?
- Trying to do the Linux build via the Docker image (like the AppVeyor CI does) just leads to a build failure with no error message somewhere in `tasks.sh`, see [here](https://github.com/DeinAlptraum/USDX/actions/runs/8976666031/job/24654126578). I have zero experience with Docker and wasn't really able to follow what it does. I tried implementing the build, including the AppImage generation, manually as you see. This generally passes the build, but for some reason the generated AppImage crashes with the error "Could not load any language file". Apparently it is looking for the language files on the host filesystem, instead of within the AppImage.  I don't know why this happens or how to fix it, and wasn't able to find any answer online either.
- I see that the AppVeyor CI also builds for Linux with 32bit. Is this a relict or do we actually want 32bit builds? I've only implemented the 64bit build so far

TL;DR this is very much a WIP except that I have no idea how to proceed with the problems I mentioned. I guess in the current state this cannot replace the existing CI, but I thought I'd still leave this here in case someone wants to pick this up or knows how to fix these problems.
Not sure if GH Actions are enabled on this repo, so [here](https://github.com/DeinAlptraum/USDX/actions/runs/8986870470)'s a link to how a build looks like currently on my fork.

### CODE REVIEW

**Comment by barbeque-squared** [comment]
Wow, this is some good progress!

> I do not know what causes this error (no, "run as admin" doesn't fix this) and googling it did not really help either. Since I have no experience with NSIS, maybe someone here knows?

I have the same issue on my own Windows machine. Regardless of how I build it, I cannot run the installer it creates. I can only run the old CI-generated installers.

> Not sure if GH Actions are enabled on this repo, so [here](https://github.com/DeinAlptraum/USDX/actions/runs/8986870470)'s a link to how a build looks like currently on my fork.

It would be good if the generated filenames included some abbreviated version information I think. Doesn't need to be the exact filename the old CI generated (I always had to remove some part of it). I'm not sure what is and isn't easy to implement, but a very naive approach would be to look at the `VERSION` file: if it ends in `dev`, put `dev-<datetime/hash/revisionnumber/whatever>` somewhere, else just put the version number in it.

> Trying to do the Linux build via the Docker image (like the AppVeyor CI does) just leads to a build failure with no error message somewhere in tasks.sh, see [here](https://github.com/DeinAlptraum/USDX/actions/runs/8976666031/job/24654126578).

There is some earlier error output on [line 1826](https://github.com/DeinAlptraum/USDX/actions/runs/8976666031/job/24654126578#step:4:1826), though I can't tell if those are expected or not. It installs these in previous steps yet they aren't available here. I don't have the time right now to look into why this is happening, but it's probably related to why the `make` command at the end errors out?

**Comment by s09bQ5** [comment]
> There is some earlier error output on [line 1826](https://github.com/DeinAlptraum/USDX/actions/runs/8976666031/job/24654126578#step:4:1826), though I can't tell if those are expected or not.

That's expected. The version check tries `cmake3` afterwards.

I think the problem is that the terminal type passed by Github in the TERM variable does not define the sgr0 capability. That would explain why it fails between "==> Building all dependencies" and "==> Cleaning prefix".

**Comment by DeinAlptraum** [comment]
Update, see the most recent build [here](https://github.com/DeinAlptraum/USDX/actions/runs/9053486579)

> I think the problem is that the terminal type passed by Github in the TERM variable does not define the sgr0 capability. That would explain why it fails between "==> Building all dependencies" and "==> Cleaning prefix".

Thanks and well spotted! I've removed all `tput` commands for now and at least we get a step further now. The build now fails because `dists/flatpak/patches/projectM-2.2.1.patch` doesn't exist. I'm not sure what's supposed to happen here tbh, I don't see any such file existing in the repository and also don't see this being generated somewhere. Do you know where this comes from?

> I have the same issue on my own Windows machine. Regardless of how I build it, I cannot run the installer it creates. I can only run the old CI-generated installers.

Through guessing and sheer luck I found out that this is an issue with versions >= 3.05 of NSIS. AppVeyor still uses 3.04 which coincidentally seems to be the latest fully working version at least for our case. Manually downgrading the runner to NSIS 3.04 thus fixed the installer. With this, only the Linux AppImage creation remains to be fixed

>It would be good if the generated filenames included some abbreviated version information I think. [...] a very naive approach would be to look at the VERSION file: if it ends in dev, put dev-<datetime/hash/revisionnumber/whatever> somewhere, else just put the version number in it.

I've appended the version (as well as `+dev-<short commit hash>` if it ends on `+dev`) to all artifact names, as well as the file if it contains only one (GHA insists on zipping even single-file artifacts), see the linked build for examples. I hope this is what you imagined

**Comment by bohning** [comment]
Judging from `binutils is keg-only, which means it was not symlinked into /usr/local, because it shadows the host toolchain.` and the fact that I do not have binutils installed and can build USDX just fine, you can probably remove binutils from the dependencies without any issues and save a few seconds.

**Comment by s09bQ5** [comment]
> The build now fails because `dists/flatpak/patches/projectM-2.2.1.patch` doesn't exist. I'm not sure what's supposed to happen here tbh, I don't see any such file existing in the repository and also don't see this being generated somewhere. Do you know where this comes from?

Fix is in #841

**Comment by s09bQ5** [comment]
> I've removed all `tput` commands for now and at least we get a step further now.

No need to do that if #842 gets merged

**Comment by DeinAlptraum** [comment]
Thanks to @s09bQ5's two changes, the build now goes through on Linux. I also tested that the AppImage works
Additionally, I added caching to the Linux build in the same way as AppVeyor did, so as long as `dl.sh` and `tasks.sh` remain unchanged, the build time should be down to about 5mins on Linux.

An example of a build with cache hit can be seen [here](https://github.com/DeinAlptraum/USDX/actions/runs/9083639630), and the build for the version I just pushed (without cache hit) can be seen [here](https://github.com/DeinAlptraum/USDX/actions/runs/9085517292)

**Review by bohning — APPROVED** [review]

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+      - name: Upload Image Artifact
+        uses: actions/upload-artifact@v4
+        with:
+          name: MAC-UltraStarDeluxe-image-${{ env.versionName }}
+          path: UltraStarDeluxe-${{ env.versionName }}.dmg
+          if-no-files-found: error
+
+  build_linux:
+    name: Build for Linux
+    runs-on: ubuntu-22.04
```
[the remark this answers is not in the document]
Comment (DeinAlptraum) [reply .github/workflows/main.yml:75]: The AppVeyor build used Ubuntu 18.04 apparently, while the oldest available one for Github Actions is Ubuntu 20.04. For now, I've opened a PR #845 to change this to 20.04 . Do you think this is acceptable in the long term?
